src/libs/kata-sys-util/src/hooks.rs

```diff
@@ -0,0 +1,519 @@
+    /// - args (array of strings, OPTIONAL) with the same semantics as IEEE Std 1003.1-2008 execv's
+    ///   argv.
+    /// - env (array of strings, OPTIONAL) with the same semantics as IEEE Std 1003.1-2008's environ.
+    /// - timeout (int, OPTIONAL) is the number of seconds before aborting the hook. If set, timeout
+    ///   MUST be greater than zero.
+    ///
+    /// The OCI spec also defines the context to invoke hooks, caller needs to take the responsibility
+    /// to setup execution context, such as namespace etc.
+    pub fn execute_hook(&mut self, hook: &oci::Hook, state: Option<oci::State>) -> Result<()> {
+        if self.get(hook) != HookState::Pending {
+            return Ok(());
+        }
+
+        fail::fail_point!("execute_hook", |_| {
+            Err(eother!("execute hook fail point injection"))
+        });
+        info!(sl!(), "execute hook {:?}", hook);
+
+        self.states.insert(hook.into(), HookState::Pending);
+
+        let mut executor = HookExecutor::new(hook)?;
+        let stdin = if state.is_some() {
+            Redirection::Pipe
+        } else {
+            Redirection::None
+        };
+        let mut popen = Popen::create(
+            &executor.args,
+            PopenConfig {
+                stdin,
+                stdout: Redirection::Pipe,
+                stderr: Redirection::Pipe,
+                executable: executor.executable.to_owned(),
+                detached: true,
+                env: Some(executor.envs.clone()),
+                ..Default::default()
+            },
+        )
+        .map_err(|e| eother!("failed to create subprocess for hook {:?}: {}", hook, e))?;
+
+        if let Some(state) = state {
+            executor.execute_with_input(&mut popen, state)?;
+        }
+        executor.execute_and_wait(&mut popen)?;
+        info!(sl!(), "hook {} finished", hook.path);
+        self.states.insert(hook.into(), HookState::Done);
+
+        Ok(())
+    }
+
+    /// Try to execute hooks and remember execution result.
+    ///
+    /// The `execute_hooks()` will be called multiple times.
+    /// It will first be called before creating the VMM when creating the sandbox, so hooks could be
+    /// used to setup environment for the VMM, such as creating tap device etc.
+    /// It will also be called during starting containers, to setup environment for those containers.
+    ///
+    /// The execution result will be recorded for each hook. Once a hook returns success, it will not
+    /// be invoked anymore.
+    pub fn execute_hooks(&mut self, hooks: &[oci::Hook], state: Option<oci::State>) -> Result<()> {
+        for hook in hooks.iter() {
+            if let Err(e) = self.execute_hook(hook, state.clone()) {
+                // Ignore error and try next hook, the caller should retry.
+                error!(sl!(), "hook {} failed: {}", hook.path, e);
+            }
+        }
+
+        Ok(())
+    }
+}
+
+struct HookExecutor<'a> {
```

Comment:
I'm rather confused by this type: why do we need  it since it stores a reference to an `oci::Hook`, which contains *the same set of fields*?!?

Reply:
It's used to prepare arguments for Popen, and helps to split a huge function into small code pieces.